xwwp.el

```diff
@@ -161,6 +161,71 @@ Interactively, URL defaults to the string looking like a url around point."
 
+;;; Adapted from EWW code to provide a DWIM style XWWP command
+(defcustom xwwp-search-prefix "https://google.com/search?q="
+  "Prefix URL to search engine."
+  :group 'xwwp
+  :type 'string)
+
+(defun xwwp--dwim-expand-url (url)
+  (setq url (string-trim url))
+  (cond ((string-match-p "\\`file:/" url))
+	    ;; Don't mangle file: URLs at all.
+        ((string-match-p "\\`ftp://" url)
+         (user-error "FTP is not supported"))
+        (t
+	     ;; Anything that starts with something that vaguely looks
+	     ;; like a protocol designator is interpreted as a full URL.
+         (if (or (string-match "\\`[A-Za-z]+:" url)
+		         ;; Also try to match "naked" URLs like
+		         ;; en.wikipedia.org/wiki/Free software
+		         (string-match "\\`[A-Za-z_]+\\.[A-Za-z._]+/" url)
+		         (and (= (length (split-string url)) 1)
+		              (or (and (not (string-match-p "\\`[\"'].*[\"']\\'" url))
+			                   (> (length (split-string url "[.:]")) 1))
+			              (string-match eww-local-regex url))))
+             (progn
+               (unless (string-match-p "\\`[a-zA-Z][-a-zA-Z0-9+.]*://" url)
+                 (setq url (concat "http://" url)))
+               ;; Some sites do not redirect final /
+               (when (string= (url-filename (url-generic-parse-url url)) "")
+                 (setq url (concat url "/"))))
+           (setq url (concat xwwp-search-prefix
+                             (mapconcat
+                              #'url-hexify-string (split-string url) "+"))))))
+  url)
+
+(defun xwwp (url &optional arg)
+  "Fetch URL and render the page.
+If the input doesn't look like an URL or a domain name, the
+word(s) will be searched for via `xwwp-search-prefix'.
+
+If called with a prefix ARG, create a new Webkit buffer instead of reusing
+the default Webkit buffer."
+  (interactive
+   (let* ((uris (eww-suggested-uris))
+	      (prompt (concat "Enter URL or keywords"
+			              (if uris (format " (default %s)" (car uris)) "")
+			              ": ")))
+     (list (read-string prompt nil 'eww-prompt-history uris)
+           (prefix-numeric-value current-prefix-arg))))
+  (setq url (xwwp--dwim-expand-url url))
+  ;; Check whether the domain only uses "Highly Restricted" Unicode
+  ;; IDNA characters.  If not, transform to punycode to indicate that
+  ;; there may be funny business going on.
+  (let ((parsed (url-generic-parse-url url)))
+    (when (url-host parsed)
+      (unless (puny-highly-restrictive-domain-p (url-host parsed))
+        (setf (url-host parsed) (puny-encode-domain (url-host parsed)))))
+    ;; When the URL is on the form "http://a/../../../g", chop off all
+    ;; the leading "/.."s.
+    (when (url-filename parsed)
+      (while (string-match "\\`/[.][.]/" (url-filename parsed))
+        (setf (url-filename parsed) (substring (url-filename parsed) 3))))
+    (setq url (url-recreate-url parsed)))
```

Comment:
Wouldn't it make sense to move this into its own function ? That way you could add a few unit test around this.